libass/ass_outline.c

```diff
@@ -133,143 +172,1008 @@ void outline_get_cbox(const ASS_Outline *outline, FT_BBox *cbox)
+    if (dx2 > -str->eps && dx2 < str->eps && dy2 > -str->eps && dy2 < str->eps) {
+        dx2 = pt[3].x - pt[1].x;
+        dy2 = pt[3].y - pt[1].y;
+        if (dx2 > -str->eps && dx2 < str->eps && dy2 > -str->eps && dy2 < str->eps)
+            return add_line(str, pt[3], dir);
+        flags ^= 4;
+    }
+
+    if (flags == 12)
+        return add_line(str, pt[3], dir);
+
+    int32_t dx1 = pt[flags >> 2].x - pt[flags & 3].x;
+    int32_t dy1 = pt[flags >> 2].y - pt[flags & 3].y;
+
+    Vector deriv[3] = {
+        { dy0 * str->yscale, -dx0 * str->xscale },
+        { dy1 * str->yscale, -dx1 * str->xscale },
+        { dy2 * str->yscale, -dx2 * str->xscale }
+    };
+    double len0 = vec_len(deriv[0]), scale0 = 1 / len0;
+    double len2 = vec_len(deriv[2]), scale2 = 1 / len2;
+    Normal normal[2] = {
+        { { deriv[0].x * scale0, deriv[0].y * scale0 }, len0 },
+        { { deriv[2].x * scale2, deriv[2].y * scale2 }, len2 }
+    };
+
+    bool first = str->contour_start;
+    if (!start_segment(str, pt[0], normal[0].v, dir))
+        return false;
+    if (!process_cubic(str, pt, deriv, normal, dir, first))
+        return false;
+    str->last_point = pt[3];
+    return true;
+}
+
+
+static bool close_contour(StrokerState *str, int dir)
+{
+    if (str->contour_start) {
+        if ((dir & 3) == 3)
+            dir = 1;
+        if (!draw_circle(str, str->last_point, dir))
+            return false;
+    } else {
+        if (!add_line(str, str->first_point, dir))
+            return false;
+        if (!start_segment(str, str->first_point, str->first_normal, dir))
+            return false;
+        if (!emit_point(str, str->first_point, str->first_normal, FT_CURVE_TAG_ON,
+                       ~str->last_skip & dir & str->first_skip))
+            return false;
+        if (str->last_normal.x != str->first_normal.x ||
+            str->last_normal.y != str->first_normal.y)
+            fix_first_point(str, str->first_point, str->last_normal,
+                           ~str->last_skip & dir & ~str->first_skip);
+        str->contour_start = true;
+    }
+    if ((dir & 1) && !outline_close_contour(str->result[0]))
+        return false;
+    if ((dir & 2) && !outline_close_contour(str->result[1]))
+        return false;
+    return true;
 }
 
+
+/*
+ * Stroke an outline glyph in x/y direction.
+ */
+bool outline_stroke(ASS_Outline *result, ASS_Outline *result1,
+                    const ASS_Outline *path, int xbord, int ybord, int eps)
+{
+    if (result1)
+        result1->n_contours = result1->n_points = 0;
+
+    int rad = FFMAX(xbord, ybord);
+    if (rad < eps) {
+        assert(result->max_contours >= path->n_contours);
+        assert(result->max_points >= path->n_points);
+        memcpy(result->contours, path->contours, sizeof(size_t) * path->n_contours);
+        memcpy(result->points, path->points, sizeof(FT_Vector) * path->n_points);
+        memcpy(result->tags, path->tags, path->n_points);
```

Comment:
So tags is 1 byte per element?

In all those cases it might be slightly more robust to use `sizeof(result->bla[0])` instead of the direct type. (I wonder whether an "array copy" macro that checks types would be good for robustness...)